### PR TITLE
feat(trie): in-memory trie node overlay

### PR DESCRIPTION
**Pull request by lakshya-sky**
Ref #7788.

### CODE REVIEW

File: crates/trie/src/trie_cursor/update.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    trie::{BranchNodeCompact, Nibbles, StoredNibbles, StoredNibblesSubKey},
+    B256,
+};
+
+/// The trie cursor factory for the trie updates.
+#[derive(Debug, Clone)]
+pub struct TrieUpdatesCursorFactory<'a, CF> {
+    cursor_factory: CF,
+    trie_updates: &'a TrieUpdatesSorted,
+}
```
Comment (Rjected) [inline crates/trie/src/trie_cursor/update.rs:14]: could we add some basic tests for this impl?

**Review by Rjected — DISMISSED** [review]
great start, we should have some tests for something like this, also interested in knowing what the log files represent / what you found when testing

**Comment by Rjected** [comment]
also looks like conflicts need to be resolved

**Comment by gakonst** [comment]
Hey @lakshya-sky, thank you so much for the initial work here. i think the design of passing around the trie updates and defining a provider that has all the updates sorted and `seek` does a `.find` to get the right one makes sense.

we'd like to take it over, as part of our Engine API revamp as it relates to how we're also modifying other things like the blockchain tree #8742

File: crates/blockchain-tree/src/chain.rs [hunk 2]
```diff
@@ -291,16 +294,17 @@ impl AppendableChain {
...
 
-        let (block_state, _) = Self::validate_and_execute(
+        let (block_state, trie_updates) = Self::validate_and_execute(
             block.clone(),
             parent_block,
             bundle_state_data,
             externals,
             block_attachment,
             block_validation_kind,
+            self.chain.trie_updates().map(|u| u.sorted()),
```
Comment (gakonst) [inline crates/blockchain-tree/src/chain.rs:304]: wonder if we can avoid the clone here somehow, maybe with `self.chain.trie_updates_mut().map(|u| u.sort_in_place())`?

would need a new function here https://github.com/paradigmxyz/reth/blob/23fc18db251381b79a17c122e28f49d6a7a6436a/crates/evm/execution-types/src/chain.rs#L82

Comment (lakshya-sky) [reply crates/blockchain-tree/src/chain.rs:304]: we want a vector for sorted updates which would need clone anyway.

**Comment by lakshya-sky** [comment]
so I've added test from the linked issue and fixed the conflicts.

**Comment by rkrasiuk** [comment]
@lakshya-sky thanks a lot for your PR. for now, i have reverted blockchain tree integration as it's in the middle of rewrite and we don't want to introduce major sensitive changes. this will be rather useful in the new version however

**Review by rkrasiuk — APPROVED** [review]

**Pull request merged**